File: lib/go/thrift/tbinary_protocol.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package thrift

import (
  "encoding/binary"
  "math"
  "strings"
  "io"
)

type TBinaryProtocol struct {
  //TProtocolBase;
  trans            TTransport
  _StrictRead      bool
  _StrictWrite     bool
  _ReadLength      int
  _CheckReadLength bool
}

type TBinaryProtocolFactory struct {
  _StrictRead  bool
  _StrictWrite bool
}

func NewTBinaryProtocolTransport(t TTransport) *TBinaryProtocol {
  return NewTBinaryProtocol(t, false, true)
}

func NewTBinaryProtocol(t TTransport, strictRead, strictWrite bool) *TBinaryProtocol {
  //return &TBinaryProtocol{TProtocolBase:TProtocolBase{trans:t}, _StrictRead:strictRead, _StrictWrite:strictWrite, _ReadLength:0, _CheckReadLength:false};
  return &TBinaryProtocol{trans: t, _StrictRead: strictRead, _StrictWrite: strictWrite, _ReadLength: 0, _CheckReadLength: false}
}

func NewTBinaryProtocolFactoryDefault() *TBinaryProtocolFactory {
  return NewTBinaryProtocolFactory(false, true)
}

func NewTBinaryProtocolFactory(strictRead, strictWrite bool) *TBinaryProtocolFactory {
  return &TBinaryProtocolFactory{_StrictRead: strictRead, _StrictWrite: strictWrite}
}

func (p *TBinaryProtocolFactory) GetProtocol(t TTransport) TProtocol {
  return NewTBinaryProtocol(t, p._StrictRead, p._StrictWrite)
}

/**
 * Writing Methods
 */

func (p *TBinaryProtocol) WriteMessageBegin(name string, typeId TMessageType, seqId int32) TProtocolException {
  if p._StrictWrite {
    version := uint32(VERSION_1) | uint32(typeId)
    e := p.WriteI32(int32(version))
    if e != nil {
      return e
    }
    e = p.WriteString(name)
    if e != nil {
      return e
    }
    e = p.WriteI32(seqId)
    return e
  } else {
    e := p.WriteString(name)
    if e != nil {
      return e
    }
    e = p.WriteByte(byte(typeId))
    if e != nil {
      return e
    }
    e = p.WriteI32(seqId)
    return e
  }
  return nil
}

func (p *TBinaryProtocol) WriteMessageEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) WriteStructBegin(name string) TProtocolException {
  return nil
}

func (p *TBinaryProtocol) WriteStructEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) WriteFieldBegin(name string, typeId TType, id int16) TProtocolException {
  e := p.WriteByte(byte(typeId))
  if e != nil {
    return e
  }
  e = p.WriteI16(id)
  return e
}

func (p *TBinaryProtocol) WriteFieldEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) WriteFieldStop() TProtocolException {
  e := p.WriteByte(STOP)
  return e
}

func (p *TBinaryProtocol) WriteMapBegin(keyType TType, valueType TType, size int) TProtocolException {
  e := p.WriteByte(byte(keyType))
  if e != nil {
    return e
  }
  e = p.WriteByte(byte(valueType))
  if e != nil {
    return e
  }
  e = p.WriteI32(int32(size))
  return e
}

func (p *TBinaryProtocol) WriteMapEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) WriteListBegin(elemType TType, size int) TProtocolException {
  e := p.WriteByte(byte(elemType))
  if e != nil {
    return e
  }
  e = p.WriteI32(int32(size))
  return e
}

func (p *TBinaryProtocol) WriteListEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) WriteSetBegin(elemType TType, size int) TProtocolException {
  e := p.WriteByte(byte(elemType))
  if e != nil {
    return e
  }
  e = p.WriteI32(int32(size))
  return e
}

func (p *TBinaryProtocol) WriteSetEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) WriteBool(value bool) TProtocolException {
  if value {
    return p.WriteByte(1)
  }
  return p.WriteByte(0)
}

func (p *TBinaryProtocol) WriteByte(value byte) TProtocolException {
  v := []byte{value}
  _, e := p.trans.Write(v)
  return NewTProtocolExceptionFromOsError(e)
}

func (p *TBinaryProtocol) WriteI16(value int16) TProtocolException {
  h := byte(0xff & (value >> 8))
  l := byte(0xff & value)
  v := []byte{h, l}
  _, e := p.trans.Write(v)
  return NewTProtocolExceptionFromOsError(e)
}

func (p *TBinaryProtocol) WriteI32(value int32) TProtocolException {
  a := byte(0xff & (value >> 24))
  b := byte(0xff & (value >> 16))
  c := byte(0xff & (value >> 8))
  d := byte(0xff & value)
  v := []byte{a, b, c, d}
  _, e := p.trans.Write(v)
  return NewTProtocolExceptionFromOsError(e)
}

func (p *TBinaryProtocol) WriteI64(value int64) TProtocolException {
  a := byte(0xff & (value >> 56))
  b := byte(0xff & (value >> 48))
  c := byte(0xff & (value >> 40))
  d := byte(0xff & (value >> 32))
  e := byte(0xff & (value >> 24))
  f := byte(0xff & (value >> 16))
  g := byte(0xff & (value >> 8))
  h := byte(0xff & value)
  v := []byte{a, b, c, d, e, f, g, h}
  _, err := p.trans.Write(v)
  return NewTProtocolExceptionFromOsError(err)
}

func (p *TBinaryProtocol) WriteDouble(value float64) TProtocolException {
  return p.WriteI64(int64(math.Float64bits(value)))
}

func (p *TBinaryProtocol) WriteString(value string) TProtocolException {
  return p.WriteBinaryFromReader(strings.NewReader(value), len(value))
}

func (p *TBinaryProtocol) WriteBinary(value []byte) TProtocolException {
  e := p.WriteI32(int32(len(value)))
  if e != nil {
    return e
  }
  _, err := p.trans.Write(value)
  return NewTProtocolExceptionFromOsError(err)
}

func (p *TBinaryProtocol) WriteBinaryFromReader(reader io.Reader, size int) TProtocolException {
  e := p.WriteI32(int32(size))
  if e != nil {
    return e
  }
  _, err := io.Copyn(p.trans, reader, int64(size))
  return NewTProtocolExceptionFromOsError(err)
}


/**
 * Reading methods
 */

func (p *TBinaryProtocol) ReadMessageBegin() (name string, typeId TMessageType, seqId int32, err TProtocolException) {
  size, e := p.ReadI32()
  if e != nil {
    return "", typeId, 0, NewTProtocolExceptionFromOsError(e)
  }
  if size < 0 {
    typeId = TMessageType(size & 0x0ff)
    version := int64(int64(size) & VERSION_MASK)
    if version != VERSION_1 {
      return name, typeId, seqId, NewTProtocolException(BAD_VERSION, "Bad version in ReadMessageBegin")
    }
    name, e = p.ReadString()
    if e != nil {
      return name, typeId, seqId, NewTProtocolExceptionFromOsError(e)
    }
    seqId, e = p.ReadI32()
    if e != nil {
      return name, typeId, seqId, NewTProtocolExceptionFromOsError(e)
    }
    return name, typeId, seqId, nil
  }
  if p._StrictRead {
    return name, typeId, seqId, NewTProtocolException(BAD_VERSION, "Missing version in ReadMessageBegin")
  }
  name, e2 := p.readStringBody(int(size))
  if e2 != nil {
    return name, typeId, seqId, e2
  }
  b, e3 := p.ReadByte()
  if e3 != nil {
    return name, typeId, seqId, e3
  }
  typeId = TMessageType(b)
  seqId, e4 := p.ReadI32()
  if e4 != nil {
    return name, typeId, seqId, e4
  }
  return name, typeId, seqId, nil
}

func (p *TBinaryProtocol) ReadMessageEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) ReadStructBegin() (name string, err TProtocolException) {
  return
}

func (p *TBinaryProtocol) ReadStructEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) ReadFieldBegin() (name string, typeId TType, seqId int16, err TProtocolException) {
  t, err := p.ReadByte()
  typeId = TType(t)
  if err != nil {
    return name, typeId, seqId, err
  }
  if t != STOP {
    seqId, err = p.ReadI16()
  }
  return name, typeId, seqId, err
}

func (p *TBinaryProtocol) ReadFieldEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) ReadMapBegin() (kType, vType TType, size int, err TProtocolException) {
  k, e := p.ReadByte()
  if e != nil {
    err = NewTProtocolExceptionFromOsError(e)
    return
  }
  kType = TType(k)
  v, e := p.ReadByte()
  if e != nil {
    err = NewTProtocolExceptionFromOsError(e)
    return
  }
  vType = TType(v)
  size32, e := p.ReadI32()
  size = int(size32)
  if e != nil {
    err = NewTProtocolExceptionFromOsError(e)
    return
  }
  return kType, vType, size, nil
}

func (p *TBinaryProtocol) ReadMapEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) ReadListBegin() (elemType TType, size int, err TProtocolException) {
  b, e := p.ReadByte()
  if e != nil {
    err = NewTProtocolExceptionFromOsError(e)
    return
  }
  elemType = TType(b)
  size32, e := p.ReadI32()
  size = int(size32)
  if e != nil {
    err = NewTProtocolExceptionFromOsError(e)
    return
  }
  return elemType, size, nil
}

func (p *TBinaryProtocol) ReadListEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) ReadSetBegin() (elemType TType, size int, err TProtocolException) {
  b, e := p.ReadByte()
  if e != nil {
    err = NewTProtocolExceptionFromOsError(e)
    return
  }
  elemType = TType(b)
  size32, e := p.ReadI32()
  size = int(size32)
  if e != nil {
    err = NewTProtocolExceptionFromOsError(e)
    return
  }
  return elemType, size, nil
}

func (p *TBinaryProtocol) ReadSetEnd() TProtocolException {
  return nil
}

func (p *TBinaryProtocol) ReadBool() (bool, TProtocolException) {
  b, e := p.ReadByte()
  v := true
  if b != 1 {
    v = false
  }
  return v, e
}

func (p *TBinaryProtocol) ReadByte() (value byte, err TProtocolException) {
  buf := []byte{0}
  err = p.readAll(buf)
  return buf[0], err
}

func (p *TBinaryProtocol) ReadI16() (value int16, err TProtocolException) {
  buf := []byte{0, 0}
  err = p.readAll(buf)
  value = int16(binary.BigEndian.Uint16(buf))
  return value, err
}

func (p *TBinaryProtocol) ReadI32() (value int32, err TProtocolException) {
  buf := []byte{0, 0, 0, 0}
  err = p.readAll(buf)
  value = int32(binary.BigEndian.Uint32(buf))
  return value, err
}

func (p *TBinaryProtocol) ReadI64() (value int64, err TProtocolException) {
  buf := []byte{0, 0, 0, 0, 0, 0, 0, 0}
  err = p.readAll(buf)
  value = int64(binary.BigEndian.Uint64(buf))
  return value, err
}

func (p *TBinaryProtocol) ReadDouble() (value float64, err TProtocolException) {
  buf := []byte{0, 0, 0, 0, 0, 0, 0, 0}
  err = p.readAll(buf)
  value = math.Float64frombits(binary.BigEndian.Uint64(buf))
  return value, err
}

func (p *TBinaryProtocol) ReadString() (value string, err TProtocolException) {
  size, e := p.ReadI32()
  if e != nil {
    return "", e
  }
  return p.readStringBody(int(size))
}

func (p *TBinaryProtocol) ReadBinary() ([]byte, TProtocolException) {
  size, e := p.ReadI32()
  if e != nil {
    return nil, e
  }
  isize := int(size)
  e = p.checkReadLength(isize)
  if e != nil {
    return nil, e
  }
  buf := make([]byte, isize)
  _, err := p.trans.ReadAll(buf)
  return buf, NewTProtocolExceptionFromOsError(err)
}

func (p *TBinaryProtocol) Flush() (err TProtocolException) {
  return NewTProtocolExceptionFromOsError(p.trans.Flush())
}

func (p *TBinaryProtocol) Skip(fieldType TType) (err TProtocolException) {
  return SkipDefaultDepth(p, fieldType)
}

func (p *TBinaryProtocol) Transport() TTransport {
  return p.trans
}

func (p *TBinaryProtocol) readAll(buf []byte) TProtocolException {
  e := p.checkReadLength(len(buf))
  if e != nil {
    return e
  }
  _, err := p.trans.ReadAll(buf)
  return NewTProtocolExceptionFromOsError(err)
}

func (p *TBinaryProtocol) setReadLength(readLength int) {
  p._ReadLength = readLength
  p._CheckReadLength = true
}

func (p *TBinaryProtocol) checkReadLength(length int) TProtocolException {
  if p._CheckReadLength {
    p._ReadLength = p._ReadLength - length
    if p._ReadLength < 0 {
      return NewTProtocolException(UNKNOWN_PROTOCOL_EXCEPTION, "Message length exceeded: "+string(length))
    }
  }
  return nil
}

func (p *TBinaryProtocol) readStringBody(size int) (value string, err TProtocolException) {
  if size < 0 {
    return "", nil
  }
  err = p.checkReadLength(size)
  if err != nil {
    return "", err
  }
  isize := int(size)
  buf := make([]byte, isize)
  _, e := p.trans.ReadAll(buf)
  return string(buf), NewTProtocolExceptionFromOsError(e)
}

File: lib/java/src/org/apache/thrift/protocol/TBinaryProtocol.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.thrift.protocol;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;

import org.apache.thrift.TException;
import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;

/**
 * Binary protocol implementation for thrift.
 *
 */
public class TBinaryProtocol extends TProtocol {
  private static final TStruct ANONYMOUS_STRUCT = new TStruct();
  private static final long NO_LENGTH_LIMIT = -1;

  protected static final int VERSION_MASK = 0xffff0000;
  protected static final int VERSION_1 = 0x80010000;

  /**
   * The maximum number of bytes to read from the transport for
   * variable-length fields (such as strings or binary) or {@link #NO_LENGTH_LIMIT} for
   * unlimited.
   */
  private final long stringLengthLimit_;

  /**
   * The maximum number of elements to read from the network for
   * containers (maps, sets, lists), or {@link #NO_LENGTH_LIMIT} for unlimited.
   */
  private final long containerLengthLimit_;

  protected boolean strictRead_;
  protected boolean strictWrite_;

  private final byte[] inoutTemp = new byte[8];

  /**
   * Factory
   */
  public static class Factory implements TProtocolFactory {
    protected long stringLengthLimit_;
    protected long containerLengthLimit_;
    protected boolean strictRead_;
    protected boolean strictWrite_;

    public Factory() {
      this(false, true);
    }

    public Factory(boolean strictRead, boolean strictWrite) {
      this(strictRead, strictWrite, NO_LENGTH_LIMIT, NO_LENGTH_LIMIT);
    }

    public Factory(long stringLengthLimit, long containerLengthLimit) {
      this(false, true, stringLengthLimit, containerLengthLimit);
    }

    public Factory(boolean strictRead, boolean strictWrite, long stringLengthLimit, long containerLengthLimit) {
      stringLengthLimit_ = stringLengthLimit;
      containerLengthLimit_ = containerLengthLimit;
      strictRead_ = strictRead;
      strictWrite_ = strictWrite;
    }

    public TProtocol getProtocol(TTransport trans) {
      return new TBinaryProtocol(trans, stringLengthLimit_, containerLengthLimit_, strictRead_, strictWrite_);
    }
  }

  /**
   * Constructor
   */
  public TBinaryProtocol(TTransport trans) {
    this(trans, false, true);
  }

  public TBinaryProtocol(TTransport trans, boolean strictRead, boolean strictWrite) {
    this(trans, NO_LENGTH_LIMIT, NO_LENGTH_LIMIT, strictRead, strictWrite);
  }

  public TBinaryProtocol(TTransport trans, long stringLengthLimit, long containerLengthLimit) {
    this(trans, stringLengthLimit, containerLengthLimit, false, true);
  }

  public TBinaryProtocol(TTransport trans, long stringLengthLimit, long containerLengthLimit, boolean strictRead, boolean strictWrite) {
    super(trans);
    stringLengthLimit_ = stringLengthLimit;
    containerLengthLimit_ = containerLengthLimit;
    strictRead_ = strictRead;
    strictWrite_ = strictWrite;
  }

  @Override
  public void writeMessageBegin(TMessage message) throws TException {
    if (strictWrite_) {
      int version = VERSION_1 | message.type;
      writeI32(version);
      writeString(message.name);
      writeI32(message.seqid);
    } else {
      writeString(message.name);
      writeByte(message.type);
      writeI32(message.seqid);
    }
  }

  @Override
  public void writeMessageEnd() throws TException {}

  @Override
  public void writeStructBegin(TStruct struct) throws TException {}

  @Override
  public void writeStructEnd() throws TException {}

  @Override
  public void writeFieldBegin(TField field) throws TException {
    writeByte(field.type);
    writeI16(field.id);
  }

  @Override
  public void writeFieldEnd() throws TException {}

  @Override
  public void writeFieldStop() throws TException {
    writeByte(TType.STOP);
  }

  @Override
  public void writeMapBegin(TMap map) throws TException {
    writeByte(map.keyType);
    writeByte(map.valueType);
    writeI32(map.size);
  }

  @Override
  public void writeMapEnd() throws TException {}

  @Override
  public void writeListBegin(TList list) throws TException {
    writeByte(list.elemType);
    writeI32(list.size);
  }

  @Override
  public void writeListEnd() throws TException {}

  @Override
  public void writeSetBegin(TSet set) throws TException {
    writeByte(set.elemType);
    writeI32(set.size);
  }

  @Override
  public void writeSetEnd() throws TException {}

  @Override
  public void writeBool(boolean b) throws TException {
    writeByte(b ? (byte)1 : (byte)0);
  }

  @Override
  public void writeByte(byte b) throws TException {
    inoutTemp[0] = b;
    trans_.write(inoutTemp, 0, 1);
  }

  @Override
  public void writeI16(short i16) throws TException {
    inoutTemp[0] = (byte)(0xff & (i16 >> 8));
    inoutTemp[1] = (byte)(0xff & (i16));
    trans_.write(inoutTemp, 0, 2);
  }

  @Override
  public void writeI32(int i32) throws TException {
    inoutTemp[0] = (byte)(0xff & (i32 >> 24));
    inoutTemp[1] = (byte)(0xff & (i32 >> 16));
    inoutTemp[2] = (byte)(0xff & (i32 >> 8));
    inoutTemp[3] = (byte)(0xff & (i32));
    trans_.write(inoutTemp, 0, 4);
  }

  @Override
  public void writeI64(long i64) throws TException {
    inoutTemp[0] = (byte)(0xff & (i64 >> 56));
    inoutTemp[1] = (byte)(0xff & (i64 >> 48));
    inoutTemp[2] = (byte)(0xff & (i64 >> 40));
    inoutTemp[3] = (byte)(0xff & (i64 >> 32));
    inoutTemp[4] = (byte)(0xff & (i64 >> 24));
    inoutTemp[5] = (byte)(0xff & (i64 >> 16));
    inoutTemp[6] = (byte)(0xff & (i64 >> 8));
    inoutTemp[7] = (byte)(0xff & (i64));
    trans_.write(inoutTemp, 0, 8);
  }

  @Override
  public void writeDouble(double dub) throws TException {
    writeI64(Double.doubleToLongBits(dub));
  }

  @Override
  public void writeString(String str) throws TException {
    byte[] dat = str.getBytes(StandardCharsets.UTF_8);
    writeI32(dat.length);
    trans_.write(dat, 0, dat.length);
  }

  @Override
  public void writeBinary(ByteBuffer bin) throws TException {
    int length = bin.limit() - bin.position();
    writeI32(length);
    trans_.write(bin.array(), bin.position() + bin.arrayOffset(), length);
  }

  /**
   * Reading methods.
   */

  @Override
  public TMessage readMessageBegin() throws TException {
    int size = readI32();
    if (size < 0) {
      int version = size & VERSION_MASK;
      if (version != VERSION_1) {
        throw new TProtocolException(TProtocolException.BAD_VERSION, "Bad version in readMessageBegin");
      }
      return new TMessage(readString(), (byte)(size & 0x000000ff), readI32());
    } else {
      if (strictRead_) {
        throw new TProtocolException(TProtocolException.BAD_VERSION, "Missing version in readMessageBegin, old client?");
      }
      return new TMessage(readStringBody(size), readByte(), readI32());
    }
  }

  @Override
  public void readMessageEnd() throws TException {}

  @Override
  public TStruct readStructBegin() throws TException {
    return ANONYMOUS_STRUCT;
  }

  @Override
  public void readStructEnd() throws TException {}

  @Override
  public TField readFieldBegin() throws TException {
    byte type = readByte();
    short id = type == TType.STOP ? 0 : readI16();
    return new TField("", type, id);
  }

  @Override
  public void readFieldEnd() throws TException {}

  @Override
  public TMap readMapBegin() throws TException {
    TMap map = new TMap(readByte(), readByte(), readI32());

    checkReadBytesAvailable(map);
    checkContainerReadLength(map.size);
    return map;
  }

  @Override
  public void readMapEnd() throws TException {}

  @Override
  public TList readListBegin() throws TException {
    TList list = new TList(readByte(), readI32());

    checkReadBytesAvailable(list);
    checkContainerReadLength(list.size);
    return list;
  }

  @Override
  public void readListEnd() throws TException {}

  @Override
  public TSet readSetBegin() throws TException {
    TSet set = new TSet(readByte(), readI32());

    checkReadBytesAvailable(set);
    checkContainerReadLength(set.size);
    return set;
  }

  @Override
  public void readSetEnd() throws TException {}

  @Override
  public boolean readBool() throws TException {
    return (readByte() == 1);
  }

  @Override
  public byte readByte() throws TException {
    if (trans_.getBytesRemainingInBuffer() >= 1) {
      byte b = trans_.getBuffer()[trans_.getBufferPosition()];
      trans_.consumeBuffer(1);
      return b;
    }
    readAll(inoutTemp, 0, 1);
    return inoutTemp[0];
  }

  @Override
  public short readI16() throws TException {
    byte[] buf = inoutTemp;
    int off = 0;

    if (trans_.getBytesRemainingInBuffer() >= 2) {
      buf = trans_.getBuffer();
      off = trans_.getBufferPosition();
      trans_.consumeBuffer(2);
    } else {
      readAll(inoutTemp, 0, 2);
    }

    return
      (short)
      (((buf[off] & 0xff) << 8) |
       ((buf[off+1] & 0xff)));
  }

  @Override
  public int readI32() throws TException {
    byte[] buf = inoutTemp;
    int off = 0;

    if (trans_.getBytesRemainingInBuffer() >= 4) {
      buf = trans_.getBuffer();
      off = trans_.getBufferPosition();
      trans_.consumeBuffer(4);
    } else {
      readAll(inoutTemp, 0, 4);
    }
    return
      ((buf[off] & 0xff) << 24) |
      ((buf[off+1] & 0xff) << 16) |
      ((buf[off+2] & 0xff) <<  8) |
      ((buf[off+3] & 0xff));
  }

  @Override
  public long readI64() throws TException {
    byte[] buf = inoutTemp;
    int off = 0;

    if (trans_.getBytesRemainingInBuffer() >= 8) {
      buf = trans_.getBuffer();
      off = trans_.getBufferPosition();
      trans_.consumeBuffer(8);
    } else {
      readAll(inoutTemp, 0, 8);
    }

    return
      ((long)(buf[off]   & 0xff) << 56) |
      ((long)(buf[off+1] & 0xff) << 48) |
      ((long)(buf[off+2] & 0xff) << 40) |
      ((long)(buf[off+3] & 0xff) << 32) |
      ((long)(buf[off+4] & 0xff) << 24) |
      ((long)(buf[off+5] & 0xff) << 16) |
      ((long)(buf[off+6] & 0xff) <<  8) |
      ((long)(buf[off+7] & 0xff));
  }

  @Override
  public double readDouble() throws TException {
    return Double.longBitsToDouble(readI64());
  }

  @Override
  public String readString() throws TException {
    int size = readI32();

    if (trans_.getBytesRemainingInBuffer() >= size) {
      String s = new String(trans_.getBuffer(), trans_.getBufferPosition(),
          size, StandardCharsets.UTF_8);
      trans_.consumeBuffer(size);
      return s;
    }

    return readStringBody(size);
  }

  public String readStringBody(int size) throws TException {
    checkStringReadLength(size);
    byte[] buf = new byte[size];
    trans_.readAll(buf, 0, size);
    return new String(buf, StandardCharsets.UTF_8);
  }

  @Override
  public ByteBuffer readBinary() throws TException {
    int size = readI32();

    checkStringReadLength(size);

    if (trans_.getBytesRemainingInBuffer() >= size) {
      ByteBuffer bb = ByteBuffer.wrap(trans_.getBuffer(), trans_.getBufferPosition(), size);
      trans_.consumeBuffer(size);
      return bb;
    }

    byte[] buf = new byte[size];
    trans_.readAll(buf, 0, size);
    return ByteBuffer.wrap(buf);
  }

  private void checkStringReadLength(int length) throws TException {
    if (length < 0) {
      throw new TProtocolException(TProtocolException.NEGATIVE_SIZE,
                                   "Negative length: " + length);
    }

    getTransport().checkReadBytesAvailable(length);

    if (stringLengthLimit_ != NO_LENGTH_LIMIT && length > stringLengthLimit_) {
      throw new TProtocolException(TProtocolException.SIZE_LIMIT,
                                   "Length exceeded max allowed: " + length);
    }
  }

  private void checkContainerReadLength(int length) throws TProtocolException {
    if (length < 0) {
      throw new TProtocolException(TProtocolException.NEGATIVE_SIZE,
                                   "Negative length: " + length);
    }
    if (containerLengthLimit_ != NO_LENGTH_LIMIT && length > containerLengthLimit_) {
      throw new TProtocolException(TProtocolException.SIZE_LIMIT,
                                   "Length exceeded max allowed: " + length);
    }
  }

  private int readAll(byte[] buf, int off, int len) throws TException {
    return trans_.readAll(buf, off, len);
  }

  /**
   *
   * Return the minimum number of bytes a type will consume on the wire
   */
  public int getMinSerializedSize(byte type) throws TTransportException {
    switch (type)
    {
      case 0: return 0; // Stop
      case 1: return 0; // Void
      case 2: return 1; // Bool sizeof(byte)
      case 3: return 1; // Byte sizeof(byte)
      case 4: return 8; // Double sizeof(double)
      case 6: return 2; // I16 sizeof(short)
      case 8: return 4; // I32 sizeof(int)
      case 10: return 8;// I64 sizeof(long)
      case 11: return 4;  // string length sizeof(int)
      case 12: return 0;  // empty struct
      case 13: return 4;  // element count Map sizeof(int)
      case 14: return 4;  // element count Set sizeof(int)
      case 15: return 4;  // element count List sizeof(int)
      default: throw new TTransportException(TTransportException.UNKNOWN, "unrecognized type code");
    }
  }
}
